Enhancing Vespa’s Embedding Management Capabilities (#314)
vespa-engine/blog: _posts/2023-06-13-enhancing-vespas-embedding-management-capabilities.md

---
layout: post
title: Enhancing Vespa’s Embedding Management Capabilities
date: '2023-06-14'
image: assets/2023-06-13-enhancing-vespas-embedding-management-capabilities/vnwayne-fan-Zqmia99hgF8-unsplash.jpg
tags: []
author: jobergum bjorncs
skipimage: true
excerpt: >
    We are thrilled to announce significant updates to Vespa’s support for inference with text embedding models
    that maps texts into vector representations.
---

![Decorative
image](/assets/2023-06-13-enhancing-vespas-embedding-management-capabilities/vnwayne-fan-Zqmia99hgF8-unsplash.jpg)
<p class="image-credit">Photo by 
<a href="https://unsplash.com/@vnwayne?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">vnwayne fan</a> 
on <a href="https://unsplash.com/photos/Zqmia99hgF8?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a></p>

We are thrilled to announce significant updates to Vespa’s support for inference with text embedding models

> that maps texts into vector representations.

Reviewer: ```suggestion
that maps texts into vector representations: General support for Huggingface models including multi-lingual embedding, embedding inference on GPUs, and new recommended models available on the Vespa Cloud model hub.
```

functionality to produce the vector embedding inside Vespa.</em></small>

## About text embedding models
Text embedding models have revolutionized natural language processing (NLP) and information retrieval tasks
by capturing the semantic meaning of unstructured text data.
Unlike traditional representations that treat words as discrete symbols,
embedding models maps text into continuous vector spaces.

![multilingual embedding model](/assets/2023-06-13-enhancing-vespas-embedding-management-capabilities/multilingual-embedding-model.png)

Embedding models trained on multilingual datasets can represent concepts across different languages enabling information retrieval across 
diverse linguistic contexts.

## Embedder Models from Huggingface
Vespa now comes with generic support for embedding models hosted on [Huggingface](https://huggingface.co/).

With the new Huggingface Embedder functionality,
developers can export embedding models from Huggingface
and import them in [ONNX](https://onnx.ai/) format in Vespa for accelerated inference close to where the data is created:

```
<container id="default" version="1.0">

>     <component id="e5" type="hugging-face-embedder">

Reviewer 2: ```suggestion
    <component id="my-embedder-id" type="hugging-face-embedder">
```

without maintaining complex fan-out relationships due to model input context length constraints.
Read more about the Huggingface embedder in the
[documentation](https://docs.vespa.ai/en/embedding.html#huggingface-embedder).

## GPU Acceleration of Embedding Models
Vespa now supports GPU acceleration of embedding model inferences.
By harnessing the power of GPUs, Vespa embedders can efficiently process large amounts of text data,
resulting in [faster response times, improved scalability, and lower cost](https://blog.vespa.ai/gpu-accelerated-ml-inference-in-vespa-cloud/).
GPU support in Vespa also unlocks using larger and more powerful embedding models
while maintaining low serving latency and cost-effectiveness.

GPU acceleration is automatically enabled in Vespa Cloud for instances where GPU(s) is available. 
Configure your stateless Vespa container cluster with a [GPU resource in services.xml](https://cloud.vespa.ai/en/reference/services#gpu).
For open-source Vespa, specify the GPU device using the
[embedder ONNX configuration](https://docs.vespa.ai/en/reference/embedding-reference.html#embedder-onnx-reference-config).  

## Vespa Model Hub Updates

> To further enrich the Vespa Cloud ecosystem,

Reviewer: ```suggestion
To make it easier to create embedding applications,
```

> we introduce new state-of-the-art text embedding models on the [Vespa Model Hub](https://cloud.vespa.ai/en/model-hub) for 

Reviewer: ```suggestion
we have added new state-of-the-art text embedding models on the [Vespa Model Hub](https://cloud.vespa.ai/en/model-hub) for 
```

```
With three lines of configuration added to the Vespa app, Vespa cloud developers can use the <code>embed</code> funcionality for 
[embedding queries](https://docs.vespa.ai/en/embedding.html#embedding-a-query-text) and [embedding document fields](https://docs.vespa.ai/en/embedding.html#embedding-a-document-field). 

Producing the embeddings closer to the Vespa storage and indexes avoids network transfer-related latency and egress costs,
which can be substantial for high-dimensional vector representations.
In addition, with Vespa Cloud’s [auto-scaling feature](https://cloud.vespa.ai/en/autoscaling),
developers do not need to worry about scaling with changes in inference traffic volume.

Vespa Cloud also allows bringing your own models using the [HuggingFace Embedder](https://docs.vespa.ai/en/embedding.html#huggingface-embedder) 
with model files submitted in the application package. In Vespa Cloud, inference with embedding models is 
automatically accelerated with GPU if the application uses [Vespa Cloud GPU instances](https://blog.vespa.ai/gpu-accelerated-ml-inference-in-vespa-cloud/).
Read more on the [Vespa Cloud model hub](https://cloud.vespa.ai/en/model-hub).

## Summary

> The improved Vespa embedding management options offer a significant leap forward in Vespa’s embedding capabilities,

Reviewer: ```suggestion
The improved Vespa embedding management options offer a significant leap in capabilities for anybody working with embeddings in online applications,
```
